contracts/programs/gummyroll/src/lib.rs

```diff
@@ -58,70 +60,196 @@ pub struct TransferAuthority<'info> {
+        // (1 << (max_depth + 1)) - 1 returns the number of nodes in the full tree
+        // The canopy size cannot exceed the size of the tree
+        if closest_power_of_2 > (1 << (max_depth + 1)) {
+            msg!(
+                "Canopy size is too large. Size: {}. Max size: {}",
+                closest_power_of_2 - 2,
+                (1 << (max_depth + 1)) - 2
+            );
+            return err!(GummyrollError::CanopyLengthMismatch);
+        }
+    } else {
+        msg!(
+            "Canopy length {} is not 2 less than a power of 2",
+            canopy.len()
+        );
+        return err!(GummyrollError::CanopyLengthMismatch);
+    }
+    // 1 is subtracted from the trailing zeros because the root is not stored in the canopy
+    Ok(closest_power_of_2.trailing_zeros() - 1)
+}
+
+fn update_canopy(
+    canopy_bytes: &mut [u8],
+    max_depth: u32,
+    change_log: Option<Box<ChangeLogEvent>>,
+) -> Result<()> {
+    // 26 is hard coded as it is the current max depth that Gummyroll supports
+    let mut empty_node_cache = Box::new([EMPTY; 26]);
+    check_canopy_bytes(canopy_bytes)?;
+    let canopy = cast_slice_mut::<u8, Node>(canopy_bytes);
+    let path_len = get_cached_path_length(canopy, max_depth)?;
+    match change_log {
+        Some(cl) => {
+            // Update the canopy from the newest change log
+            for path_node in cl.path.iter().rev().skip(1).take(path_len as usize) {
+                // node_idx - 2 maps to the canopy index
+                canopy[(path_node.index - 2) as usize] = path_node.node;
+            }
+        }
+        None => {
+            // Initialize the canopy with empty nodes
+            for (i, node) in canopy.iter_mut().enumerate() {
+                let node_idx = (i + 2) as u32;
+                let level = max_depth - (31 - node_idx.leading_zeros());
+                *node = empty_node_cached::<26>(level, &mut empty_node_cache);
+            }
+        }
+    }
+    Ok(())
+}
+
+fn fill_in_proof_from_canopy(
+    canopy_bytes: &mut [u8],
+    max_depth: u32,
+    index: u32,
+    proof: &mut Vec<Node>,
+) -> Result<()> {
+    check_canopy_bytes(canopy_bytes)?;
+    let canopy = cast_slice_mut::<u8, Node>(canopy_bytes);
+    let path_len = get_cached_path_length(canopy, max_depth)?;
+    // We want to compute the node index (w.r.t. the canopy) where the current path
+    // intersects the leaves of the canopy
+    let mut node_idx = ((1 << max_depth) + index) >> (max_depth - path_len);
+    let mut inferred_nodes = vec![];
+    while node_idx > 1 {
+        // node_idx - 2 maps to the canopy index
+        let shifted_index = node_idx as usize - 2;
+        if shifted_index % 2 == 0 {
+            inferred_nodes.push(canopy[shifted_index + 1])
+        } else {
+            inferred_nodes.push(canopy[shifted_index - 1])
+        }
+        node_idx >>= 1;
+    }
+    proof.extend(inferred_nodes.iter());
```

Comment:
The correct behavior here is probably to only extend if the proof node is not already provided. Will add the change in soon